### PR TITLE
feat(config): introduce config version updater

### PR DESCRIPTION
**Pull request by MatusKysel**
#1402

### CODE REVIEW

**Review by gaoqi8649 — APPROVED** [review]

File: libraries/cli/src/config_updater.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    old_conf["network_listen_ip"] = old_conf["network_address"];
+    old_conf.removeMember("network_address");
+  } else if (old_conf["network_listen_ip"].isNull()) {
+    old_conf["network_listen_ip"] = new_conf["network_listen_ip"];
+  }
+}
+//// NetworkIPChange
+
+ConfigUpdater::ConfigUpdater(int network_id) {
+  new_conf_ = Tools::generateConfig((Config::NetworkIdType)network_id);
```
Comment (JakubFornadel) [inline libraries/cli/src/config_updater.cpp:25]: static_cast

File: libraries/cli/src/config_updater.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  // Regiser changes that should applay
```
Comment (JakubFornadel) [inline libraries/cli/src/config_updater.cpp:26]: typo applay

File: CMakeLists.txt [hunk 3]
```diff
@@ -14,6 +14,8 @@ set(TARAXA_DB_MAJOR_VERSION 2)
 # Minor version should be modified when changes to the database are made in the tables that can be rebuilt from the
 # basic tables
 set(TARAXA_DB_MINOR_VERSION 3)
+# If config params are changed we should increase the version
+set(TARAXA_CONFIG_VERSION 1)
```
Comment (JakubFornadel) [inline CMakeLists.txt:18]: why dont we use node version instead of this ?

File: libraries/cli/src/config.cpp [hunk 4]
```diff
@@ -168,6 +169,12 @@ Config::Config(int argc, const char* argv[]) {
       fs::create_directories(data_dir);
     }
 
+    // if there is not chain id it is a test
```
Comment (JakubFornadel) [inline libraries/cli/src/config.cpp:172]: Why do we create config updater only if this condition pass ? Shouldnt it be created every time ?

File: libraries/cli/include/cli/config_updater.hpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+class ConfigUpdater {
+ public:
+  ConfigUpdater(int network_id);
+  ~ConfigUpdater() = default;
+
+  void UpdateConfig(const std::string& config_name, Json::Value& old_conf);
+
+ private:
+  Json::Value new_conf_;
+  std::vector<std::unique_ptr<ConfigChange>> config_changes_;
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/config_updater.hpp:36]: Do we need this while boilerplate around ConfigChange class and derived classes.... ? Wouldnt be somehting like this  simpler and achieve the same thing ?

struct ConfigChange {
  VersionType version;
  std::function<void(const std::string& config_name, Json::Value& old_conf)> apply;
}

std::vector<ConfigChange> config_changes_;

private:
void registerConfigChanges() {
  // register all config changes functions

  // IP config change
  ConfigChange networkIPChange(1.0.0, [](Json::Value& old_conf, const Json::Value& new_conf){ code... });
 config_changes_.push_back(networkIPChange)
}

File: CMakeLists.txt (hunk 3 above)
Comment (MatusKysel) [reply CMakeLists.txt:18]: changing of node version does not mean, changing the config. But I guess we can do it like that

File: libraries/cli/src/config.cpp (hunk 4 above)
Comment (MatusKysel) [reply libraries/cli/src/config.cpp:172]: read the comment above you line. As test do not have chain id, I can not apply update

**Review by JakubFornadel — APPROVED** [review]

**Pull request merged**